core/src/main/scala/org/apache/spark/crypto/CryptoInputStream.scala

```diff
@@ -0,0 +1,439 @@
+      tmpBuf = new Array[Byte](bufferSize)
+    }
+    tmpBuf
+  }
+
+  /**
+   * Do the decryption using inBuffer as input and outBuffer as output.
+   * Upon return, inBuffer is cleared the decrypted data starts at
+   * outBuffer.position() and ends at outBuffer.limit()
+   */
+  def decrypt(decryptor: Decryptor, inBuffer: ByteBuffer, outBuffer: ByteBuffer, padding: Byte) {
+    Preconditions.checkState(inBuffer.position >= padding)
+    if (inBuffer.position != padding) {
+      inBuffer.flip
+      outBuffer.clear
+      decryptor.decrypt(inBuffer, outBuffer)
+      inBuffer.clear
+      outBuffer.flip
+      if (padding > 0) {
+        outBuffer.position(padding)
+      }
+    }
+  }
+
+  /**
+   * This method is executed immediately after decryption. Check whether
+   * decryptor should be updated and recalculate padding if needed.
+   */
+  def afterDecryption(decryptor: Decryptor, inBuffer: ByteBuffer, position: Long,
+                      iv: Array[Byte]): Byte = {
+    var padding: Byte = 0
+    if (decryptor.isContextReset) {
+      updateDecryptor(decryptor, position, iv)
+      padding = getPadding(position)
+      inBuffer.position(padding)
+    }
+    padding
+  }
+
+  def getCounter(position: Long): Long = {
+    position / codec.getCipherSuite.algoBlockSize
+  }
+
+  def getPadding(position: Long): Byte = {
+    (position % codec.getCipherSuite.algoBlockSize).asInstanceOf[Byte]
+  }
+
+  /** Calculate the counter and iv, update the decryptor. */
+  def updateDecryptor(decryptor: Decryptor, position: Long, iv: Array[Byte]) {
+    val counter: Long = getCounter(position)
+    codec.calculateIV(initIV, counter, iv)
+    decryptor.init(key, iv)
+  }
+
+  /**
+   * Reset the underlying stream offset clear {@link #inBuffer} and
+   * {@link #outBuffer}. This Typically happens during {@link #seek(long)}
+   * or {@link #skip(long)}.
+   */
+  def resetStreamOffset(offset: Long) {
+    streamOffset = offset
+    inBuffer.clear
+    outBuffer.clear
+    outBuffer.limit(0)
+    updateDecryptor(decryptor, offset, iv)
+    padding = getPadding(offset)
+    inBuffer.position(padding)
+  }
+
+  override def close {
+    if (!closed) {
+      super.close
+      freeBuffers
+      closed = true
+    }
+  }
+
+  /** Skip n bytes */
+  override def skip(nVal: Long): Long = {
+    var n: Long = nVal
+    Preconditions.checkArgument(n >= 0, "Negative skip length.", new Array[String](1))
+    checkStream
+    if (n == 0) {
+      0
+    }
+    else if (n <= outBuffer.remaining) {
+      val pos: Int = outBuffer.position + n.asInstanceOf[Int]
+      outBuffer.position(pos)
+      n
+    }
+    else {
+      n -= outBuffer.remaining
+      var skipped: Long = in.skip(n)
+      if (skipped < 0) {
+        skipped = 0
+      }
+      val pos: Long = streamOffset + skipped
+      skipped += outBuffer.remaining
+      resetStreamOffset(pos)
+      skipped
+    }
+  }
+
+  /** ByteBuffer read. */
+  override def read(buf: ByteBuffer): Int = {
```

Comment:
So, this method is really hard to read. It's a maze of `ifs` and `elses` and it's hard to figure out what's being returned and what's actually being done.

Can you rewrite it so that you use explicit `return`s instead and avoid the nested conditionals?